### PR TITLE
[Feature] Add QUALIFY for window_function(ROW_NUMBER/RANK/DESN_RANK)

### PR DESCRIPTION
**Pull request by Pslydhh**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Problem Summary(Required) ：

We could use **QUALIFY** with window function(ROW_NUMBER/RANK/DENSE_RANK) to filter data in one SELECT.
```
SELECT STATISTICAL_DATE FROM window_table_bucket_10 QUALIFY ROW_NUMBER() OVER(PARTITION BY PROVINCE_CODE order by TOTAL_SCORE) = 1;
```
is equal with:
```
SELECT STATISTICAL_DATE FROM (SELECT STATISTICAL_DATE, ROW_NUMBER() OVER(PARTITION BY PROVINCE_CODE order by TOTAL_SCORE) as score from window_table_bucket_10) t where t.score = 1;
```
## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 28 / 29 (96.55%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|26|27|96.30%|[3181]|
|:large_blue_circle:|com/starrocks/sql/ast/SelectListItem.java|2|2|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4 [hunk 1]
```diff
@@ -1476,6 +1476,7 @@ limitElement
 querySpecification
     : SELECT setVarHint* setQuantifier? selectItem (',' selectItem)*
       fromClause
+      (QUALIFY qualifyFunction=selectItem '=' limit=INTEGER_VALUE)?
```
[the remark this answers is not in the document]
Comment (HangyuanLiu) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:1479]: The grammar file here cannot be modified in this way. By definition, QUALIFY should be the outermost layer of a sql. If it is so simple to add a keyword, how to report an error to the sql of "select tc from tall qualify dense_rank() OVER(PARTITION by ta order by tg) = 1 where c1 = 2"?

[the remark this answers is not in the document]
Comment (Pslydhh) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4]: > The grammar file here cannot be modified in this way. By definition, QUALIFY should be the outermost layer of a sql. If it is so simple to add a keyword, how to report an error to the sql of "select tc from tall qualify dense_rank() OVER(PARTITION by ta order by tg) = 1 where c1 = 2"?

How about use sql with QUALIFY as independent case(without where/groupby/having)?

[the remark this answers is not in the document]
Comment (Pslydhh) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4]: > you only support `= 1/2/3/4`?

Yes, For window function(row_number/rank/dense_rank), just support INTEGER_VALUE now, maybe we could improve It to use Expr  at future.

[the remark this answers is not in the document]
Comment (Pslydhh) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4]: > you only support `= 1/2/3/4`?

Have add other comparisonOperator.

**Comment by Pslydhh** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 34 / 34 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|32|32|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/SelectListItem.java|2|2|100.00%|[]|

**Review by HangyuanLiu — APPROVED** [review]

**Review by murphyatwork — APPROVED** [review]

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Pull request merged**